### PR TITLE
Fix latex problems

### PR DESCRIPTION
**Pull request by Fatalll**

### CODE REVIEW

File: src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt [hunk 1]
```diff
@@ -1,22 +1,40 @@
...
+                && PsiTreeUtil.findFirstParent(element) { it is LatexOpenGroup }?.let { latexOpenGroup ->
+            PsiTreeUtil.findFirstParent(element) { it is LatexGroup }?.let { latexOpenGroup.isAncestorOf(it) } ?: false
+        } ?: true
     }
 
     override fun check(element: PsiElement): Set<Typo> {
         require(element is LatexNormalText) { "Got non LatexNormalText in LatexSupport" }
-        return GrammarChecker.default.check(element.filterForTokens<PsiElement>(NORMAL_TEXT_WORD)).toSet()
+        return GrammarChecker.default.check(
+                element.filterForTokens<PsiElement>(NORMAL_TEXT_WORD)
```
Comment (TanVD) [inline src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt]: Unneeded generic

File: src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt [hunk 2]
```diff
@@ -1,22 +1,40 @@
...
 import tanvd.grazi.ide.language.LanguageSupport
 import tanvd.grazi.utils.filterForTokens
 
 class LatexSupport : LanguageSupport() {
+    private val ignoredCategories = listOf(Typo.Category.CASING, Typo.Category.PUNCTUATION, Typo.Category.GRAMMAR)
+
     override fun isRelevant(element: PsiElement): Boolean {
         return element is LatexNormalText && element.parents.none { it is LatexMathEnvironment }
+                && PsiTreeUtil.findFirstParent(element) { it is LatexOpenGroup }?.let { latexOpenGroup ->
+            PsiTreeUtil.findFirstParent(element) { it is LatexGroup }?.let { latexOpenGroup.isAncestorOf(it) } ?: false
```
Comment (TanVD) [inline src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt]: What should mean this situation, when `LatexOpenGroup` `isAncestorOf` `LatexGroup`? Move it to function and name accordingly :)

File: src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt (hunk 1 above)
Comment (TanVD) [reply src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt]: And should we really filter for NORMAL_TEXT_WORD? Why not to check the whole LatexNormalText?

File: src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt [hunk 3]
```diff
@@ -1,22 +1,40 @@
...
     }
 
     override fun check(element: PsiElement): Set<Typo> {
         require(element is LatexNormalText) { "Got non LatexNormalText in LatexSupport" }
-        return GrammarChecker.default.check(element.filterForTokens<PsiElement>(NORMAL_TEXT_WORD)).toSet()
+        return GrammarChecker.default.check(
+                element.filterForTokens<PsiElement>(NORMAL_TEXT_WORD)
+        ).filterNot { typo ->
+            typo.info.category in ignoredCategories &&
+                    typo.location.element?.let {
```
Comment (TanVD) [inline src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt]: Move condition to separate function with name explaining what it means

File: src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt (hunk 2 above)
Comment (TanVD) [reply src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt]: (see as an example implementation of Markdown with functions like `isParagraph`, `isHeader` and so on

File: src/test/resources/ide/language/latex/Example.tex [hunk 4]
```diff
@@ -4,11 +4,42 @@
...
+
+    Our goal was to minimize the code length, it <warning descr="IT_VBZ">are</warning> natural to compare the result with other methods.
+    We have used arithmetic coding (AC) as (categories) <warning descr="EN_A_VS_AN">a</warning> entropy encoder.
+    <warning descr="UNLIKELY_OPENING_PUNCTUATION">,</warning> bad start of the sentence.
+
+    \begin{itemize}
+        \item are not bad
+        \item two
+        \item three
+    \end{itemize}
```
Comment (TanVD) [inline src/test/resources/ide/language/latex/Example.tex:23]: I would add here one more paragraph of text :)

File: src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt (hunk 1 above)
Comment (Fatalll) [reply src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt]: Generic is needed

Comment (Fatalll) [reply src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt]: Well, I need a position of typos, so if use whole LatexNormalText, each typo will be at zero position.

File: src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt (hunk 2 above)
Comment (Fatalll) [reply src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt]: Done

File: src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt (hunk 3 above)
Comment (Fatalll) [reply src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt]: Fixed

File: src/test/resources/ide/language/latex/Example.tex (hunk 4 above)
Comment (Fatalll) [reply src/test/resources/ide/language/latex/Example.tex:23]: Added

File: src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt (hunk 1 above)
Comment (Fatalll) [reply src/main/kotlin/tanvd/grazi/ide/language/latex/LatexSupport.kt]: Fixed

File: src/main/kotlin/tanvd/grazi/grammar/Typo.kt [hunk 5]
```diff
@@ -92,4 +97,22 @@ data class Typo(val location: Location, val info: Info, val fixes: List<String>
             }
         }
     }
+
+    fun isAtStart(): Boolean {
```
Comment (TanVD) [inline src/main/kotlin/tanvd/grazi/grammar/Typo.kt]: better move to location :)

Comment (TanVD) [reply src/main/kotlin/tanvd/grazi/grammar/Typo.kt]: And add parameter `skipWhitespace = true`. In soma cases we may not want to skip whitespaces

Comment (Fatalll) [reply src/main/kotlin/tanvd/grazi/grammar/Typo.kt]: Done

**Review by TanVD — APPROVED** [review]

**Pull request merged**